proto-lens/src/Data/ProtoLens/Encoding/Growing.hs

```diff
@@ -0,0 +1,59 @@
+-- | A mutable vector that grows in size.
+module Data.ProtoLens.Encoding.Growing (
+    Growing,
+    new,
+    append,
+    unsafeFreeze,
+    RealWorld,
+    ) where
+
+import Control.Monad.Primitive (PrimMonad, PrimState, RealWorld)
+
+import qualified Data.Vector.Generic as V
+import qualified Data.Vector.Generic.Mutable as MV
+
+-- | A mutable vector which can increase in capacity.
+data Growing v s a = Growing
+    {-# UNPACK #-} !Int
+        -- The number of elements in the mutable vector
+        -- that have already been set.
+    !(V.Mutable v s a)
+        -- TODOs for efficiency:
+        -- - Try unpacking this.  It's difficult as-is because
+        --   V.Mutable is a type function.
+        -- - MVectors support slicing, but we're not using that
+        --   functionality, so we're passing around an extra unnecessary
+        --   Int.
+
+-- | Create a new empty growing vector.
+new :: (PrimMonad m, V.Vector v a) => m (Growing v (PrimState m) a)
+new = Growing 0 <$> MV.new 0
+
+-- | Unsafely convert a growing vector to an immutable one without
+-- copying.  After this call, you may not use the growing vector
+-- nor any other growing vectors that were used to produce this one.
+unsafeFreeze
+    :: (PrimMonad m, V.Vector v a)
+    => Growing v (PrimState m) a -> m (v a)
+unsafeFreeze (Growing len m)
+    = V.unsafeFreeze (MV.take len m)
```

Comment:
Move up to conserve some vertical space?

Reply:
Done.